### PR TITLE
support for `ext_conservative_rasterization`

### PR DESCRIPTION
**Pull request by LukeP0WERS**
added a conservative rasterization state to the graphics pipeline and the dynamic states for it. i have tested it with my own project and everything seems to be working.

Changelog:
```markdown
### Additions
- Support for the `ext_conservative_rasterization` extension.
- Support for the `extended_dynamic_state3_conservative_rasterization_mode` device feature.
- Support for the `extended_dynamic_state3_extra_primitive_overestimation_size` device feature.
```

### CODE REVIEW

File: vulkano/src/command_buffer/commands/pipeline.rs [hunk 1]
```diff
@@ -3216,6 +3216,40 @@ impl RecordingCommandBuffer {
...
+                        return Err(Box::new(ValidationError {
+                            problem: format!(
+                                "the currently bound graphics pipeline requires the \
+                                `DynamicState::{:?}` dynamic state, but \
+                                this state was either not set, or it was overwritten by a \
+                                more recent `bind_pipeline_graphics` command",
+                                dynamic_state
+                            )
+                            .into(),
+                            ..Default::default()
```
Comment (Rua) [inline vulkano/src/command_buffer/commands/pipeline.rs:3230]:
```suggestion
                            vuids: vuids!(vuid_type, "None-07631")
                            ..Default::default()
```

File: vulkano/src/command_buffer/commands/pipeline.rs [hunk 2]
```diff
@@ -3216,6 +3216,40 @@ impl RecordingCommandBuffer {
...
```
Comment (Rua) [inline vulkano/src/command_buffer/commands/pipeline.rs:3249]:
```suggestion
                            vuids: vuids!(vuid_type, "None-07632")
                            ..Default::default()
```

**Comment by Rua** [comment]
A few VUIDs are not validated yet:
- VUID-VkGraphicsPipelineCreateInfo-conservativePointAndLineRasterization-08892
- VUID-vkCmdDraw-conservativePointAndLineRasterization-07499

**Comment by LukeP0WERS** [comment]
not sure how to do VUID-vkCmdDraw*-conservativePointAndLineRasterization-07499

File: vulkano/src/pipeline/graphics/mod.rs [hunk 3]
```diff
@@ -2572,6 +2573,130 @@ impl GraphicsPipelineCreateInfo {
...
         }
 
+        if match device
+            .physical_device()
+            .properties()
+            .conservative_point_and_line_rasterization
+        {
+            Some(b) => !b,
+            None => true,
+        } {
```
Comment (Rua) [inline vulkano/src/pipeline/graphics/mod.rs:2583]:
```suggestion
        let properties = device.physical_device().properties();

        if matches!(conservative_rasterization_state.mode, ConservativeRasterizationMode::Disabled)
            && !properties.conservative_point_and_line_rasterization.unwrap_or(false)
        {
```

And then all the matches for `ConservativeRasterizationMode::Disabled` further below are no longer needed.

File: vulkano/src/pipeline/graphics/mod.rs [hunk 4]
```diff
@@ -2572,6 +2573,130 @@ impl GraphicsPipelineCreateInfo {
...
+            ) {
+                if match conservative_rasterization_state.mode {
+                    ConservativeRasterizationMode::Disabled => false,
+                    _ => true,
+                } && match input_assembly_state.topology {
+                    PrimitiveTopology::PointList => true,
+                    PrimitiveTopology::LineList => true,
+                    PrimitiveTopology::LineStrip => true,
+                    _ => false,
+                } && (!dynamic_state.contains(&DynamicState::PrimitiveTopology)
```
Comment (Rua) [inline vulkano/src/pipeline/graphics/mod.rs:2597]:
```suggestion
                } && matches!(
                    input_assembly_state.topology,
                    PrimitiveTopology::PointList
                    | PrimitiveTopology::LineList
                    | PrimitiveTopology::LineStrip
                ) && (!dynamic_state.contains(&DynamicState::PrimitiveTopology)
```

File: vulkano/src/pipeline/graphics/mod.rs [hunk 5]
```diff
@@ -2572,6 +2573,130 @@ impl GraphicsPipelineCreateInfo {
...
+                        ..Default::default()
+                    }));
+                }
+            }
+
+            if let (Some(mesh_stage), Some(_), Some(conservative_rasterization_state)) = (
+                mesh_stage,
+                input_assembly_state,
+                conservative_rasterization_state,
+            ) {
```
Comment (Rua) [inline vulkano/src/pipeline/graphics/mod.rs:2662]:
```suggestion
            if let (Some(mesh_stage), Some(conservative_rasterization_state)) = (
                mesh_stage,
                conservative_rasterization_state,
            ) {
```

Pipelines with mesh shaders have no input assembly state.

File: vulkano/src/pipeline/graphics/mod.rs [hunk 6]
```diff
@@ -2572,6 +2573,130 @@ impl GraphicsPipelineCreateInfo {
...
+                                break;
+                            }
+                            ExecutionMode::OutputLineStrip => {
+                                invalid_output = true;
+                                break;
+                            }
+                            _ => {}
+                        }
+                    }
+                }
```
Comment (Rua) [inline vulkano/src/pipeline/graphics/mod.rs:2641]:
```suggestion
                let invalid_output = entry_point_function
                    .execution_modes()
                    .any(|instruction| matches!(
                        instruction,
                        Instruction::ExecutionMode {
                            mode: ExecutionMode::OutputPoints | ExecutionMode::OutputLineStrip,
                            ..
                        },
                    ));
```

**Pull request merged**